dayforce_client/client.py

```diff
@@ -105,17 +110,83 @@ class DayforceSFTP(object):
     disable_host_key_checking: bool = attr.ib(default=False)
     cnopts: pysftp.CnOpts = attr.ib(init=False)
 
+    _sftp: Optional[pysftp.Connection] = attr.ib(default=None)
+    _sftp_live: bool = attr.ib(default=False)
+
     def __attrs_post_init__(self):
         self.cnopts = pysftp.CnOpts()
         if self.disable_host_key_checking is True:
             self.cnopts.hostkeys = None
         elif self.host_key is not None:
             key_b = self.host_key.encode()
             key = paramiko.RSAKey(data=base64.decodebytes(key_b))
-            self.cnopts.hostkeys.add(self.hostname, 'ssh-rsa', key)
+            self.cnopts.hostkeys.add(self.hostname, "ssh-rsa", key)
+        else:
+            raise RuntimeError("disable_host_key_checking or host_key must be set")
+
+    def _connect(self):
+        """Establish the SFTP connection."""
+        if not self._sftp_live:
+            self._sftp = pysftp.Connection(
+                host=self.hostname,
+                username=self.username,
+                password=self.password,
+                port=self.port,
+                cnopts=self.cnopts,
+            )
+            self._sftp_live = True
+
+    def _disconnect(self):
+        """Close the SFTP connection."""
+        if self._sftp_live:
+            self._sftp.close()
+            self._sftp_live = False
+
+    @contextmanager
+    def connect(self):
+        self._connect()
+        try:
+            yield self
+        except Exception as e:
+            self._disconnect()
+            raise e
+
+    """Upload a batch import file and return a token for status checking"""
+
+    def put_import(self, filename: str, type: str) -> str:
+        self._connect()
```

Comment:
@bobzoller If I'm reading this correctly, it seems like if a client were to call `put_import()` outside of a context manager, we would open the connection for them but we would **not** close the connection. Do you think it's desirable to also add logic to close the connection as well?

Reply:
yeah good q.  honestly I was cribbing off of pysftp's own internal workings IIRC.  I think I'd be more inclined to raise if you tried to do something that required a connection, rather than making the connect logic more complex.  FWIW, I don't know that there's anything negative that happens if you don't actually disconnect (which may be why pysftp is written this way too).

do you feel strongly enough one way or the other that I should rework it in this PR?

Reply:
reworked.  now the first call to `connect` will manage disconnecting, but subsequent ones won't.